dart/lib/src/builder.dart

```diff
@@ -0,0 +1,623 @@
+  }
+
+  void _integrityCheckOnKeyAddition() {
+    if(_finished) {
+      throw Exception('Adding values after finish is prohibited');
+    }
+    if(_stackPointers.isEmpty || _stackPointers.last.isVector) {
+      throw Exception('Adding key before staring a map is prohibited');
+    }
+  }
+
+  void _finish() {
+    if (_stack.length != 1) {
+      throw Exception('Stack has to be exactly 1');
+    }
+    var value = _stack[0];
+    var byteWidth = _align(value.elementWidth(_offset, 0));
+    _writeStackValue(value, byteWidth);
+    _writeInt(value.storedPackedType(), 1);
+    _writeInt(byteWidth, 1);
+    _finished = true;
+  }
+  
+  _StackValue _createVector(int start, int vecLength, int step, [_StackValue keys]) {
+    var bitWidth = BitWidthUtil.width(vecLength);
+    var prefixElements = 1;
+    if (keys != null) {
+      var elemWidth = keys.elementWidth(_offset, 0);
+      if (elemWidth.index > bitWidth.index) {
+        bitWidth = elemWidth;
+      }
+      prefixElements += 2;
+    }
+    var vectorType = ValueType.Key;
+    var typed = keys == null;
+    for (var i = start; i < _stack.length; i+=step) {
+      var elemWidth = _stack[i].elementWidth(_offset, i + prefixElements);
+      if (elemWidth.index > bitWidth.index) {
+        bitWidth = elemWidth;
+      }
+      if (i == start) {
+        vectorType = _stack[i].type;
+        typed &= ValueTypeUtils.isTypedVectorElement(vectorType);
+      } else {
+        if (vectorType != _stack[i].type) {
+          typed = false;
+        }
+      }
+    }
+    var byteWidth = _align(bitWidth);
+    var fix = typed & ValueTypeUtils.isNumber(vectorType) && vecLength >= 2 && vecLength <= 4;
+    if (keys != null) {
+      _writeStackValue(keys, byteWidth);
+      _writeInt(1 << keys.width.index, byteWidth);
+    }
+    if (fix == false) {
+      _writeInt(vecLength, byteWidth);
+    }
+    var vecOffset = _offset;
+    for (var i = start; i < _stack.length; i+=step) {
+      _writeStackValue(_stack[i], byteWidth);
+    }
+    if (typed == false) {
+      for (var i = start; i < _stack.length; i+=step) {
+        _writeInt(_stack[i].storedPackedType(), 1);
+      }
+    }
+    if (keys != null) {
+      return _StackValue.WithOffset(vecOffset, ValueType.Map, bitWidth);
+    }
+    if(typed) {
+      var vType = ValueTypeUtils.toTypedVector(vectorType, fix ? vecLength : 0);
+      return _StackValue.WithOffset(vecOffset, vType, bitWidth);
+    }
+    return _StackValue.WithOffset(vecOffset, ValueType.Vector, bitWidth);
+  }
+
+  void _endVector(_StackPointer pointer) {
+    var vecLength = _stack.length - pointer.stackPosition;
+    var vec = _createVector(pointer.stackPosition, vecLength, 1);
+    _stack.removeRange(pointer.stackPosition, _stack.length);
+    _stack.add(vec);
+  }
+
+  void _sortKeysAndEndMap(_StackPointer pointer) {
+    if (((_stack.length - pointer.stackPosition) & 1) == 1) {
+      throw Exception('The stack needs to hold key value pairs (even number of elements)');
```

Comment:
`StateError`.

The error message should also be cleaned up a bit. What does this mean to me as a developer? How should I change my program or schema to fix this?